src/const-safety.md

```diff
@@ -0,0 +1,171 @@
+is defined as follows:
+
+* Integer and floating point types are const-safe if they are a `Scalar::Bits`.
+  This makes sure that we can run `%` and other operations without violating
+  const safety.  In particular, the value must *not* be uninitialized.
+* References are const-safe if they are `Scalar::Ptr` into allocated memory, and
+  the data stored there is const-safe.  (Technically, we would also like to
+  require `&mut` to be unique and the memory behind a `&` to not be mutable unless there is an
+  `UnsafeCell`, but it seems infeasible to check that.)  For wide pointers, the
+  length of a slice must be a valid `usize` and the vtable of a `dyn Trait` must
+  be a valid vtable.
+* `bool` is const-safe if it is `Scalar::Bits` with a value of `0` or `1`.
+* `char` is const-safe if it is a valid unicode codepoint.
+* `()` is always const-safe.
+* `!` is never const-safe.
+* `dyn Trait` is const-safe if the value is const-safe at the type indicated by
+  the vtable.
+* Function pointers are const-safe if they point to an actual function.  A
+  `const fn` pointer (when/if we have those) must point to a `const fn`.
+* Raw pointers are const safe if they are either `Scalar::Bits` or are not dangling
+* Aggregates must not contain dangling `Scalar::Ptr` (e.g. in padding or unions) and additionally
+  follow the following aggregate-specific rules:
+    * Tuples, structs, arrays and slices are const-safe if all their fields are
+    const-safe.
+    * Enums are const-safe if they have a valid discriminant and the fields of the
+    active variant are const-safe.
+    * Unions are always const-safe; the data does not matter.
+
+
+For example:
+
+```rust
+static S: i32 = 0;
+const BAD: usize = unsafe { &S as *const i32 as usize };
+```
+
+Here, `S` is const-safe because `0` is a `Scalar::Bits`.  However, `BAD` is *not* const-safe because it is a `Scalar::Ptr`.
+Also in
+
+```rust
+static X: i32 = 42;
+static GOOD: *const i32 = &X;
+static BAD: *const i32 = { let x = 42; &x as *const i32 };
+```
+
+`BAD` is not const-safe, because it points to the local variable `x` which will be deallocated, while
+`GOOD` is const-safe, because it points to some memory that lives as long as the static itself.
+
+
+## Const safety check on code
+
+The purpose of the const safety check on code is to prohibit construction of
+non-const-safe values in safe code.  We can allow *almost* all runtime-safe operations,
+except for unconst operations -- which are all related to raw pointers:
+
+* Comparing raw pointers for (in)equality or order,
+* converting raw pointers to integers,
+* inspecting the bits of raw pointers (e.g. by hashing them).
+
+
+These operations are unconst because a `const fn` that
+(when called with the same arguments at runtime and compile-time)
+
+  * fails to run at compile time when it works perfectly fine at runtime
+  * produces a different result at compile time than at runtime
+
+is undefined behavior.
```

Comment:
That would be a new kind of UB to be [added to the reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html), and to be checked by Miri. I think the latter will be really hard to do so I am not sure if it's worth it.

Reply:
Yea, this is from before the `guaranteed_*` PR where I ended up agreeing that making it UB likely isn't worth it.